### PR TITLE
Add documentation for FeatureAlphaDropout

### LINKED ISSUE #37436: feature_alpha_dropout is not documented
https://pytorch.org/docs/master/search.html?q=feature_alpha_dropout&check_keywords=yes&area=default#

### PR DESCRIPTION
**Pull request by choidongyeon**
These changes add documentation for FeatureAlphaDropout, based on a need raised in an issue by @SsnL (Issue #9886). Fixes https://github.com/pytorch/pytorch/issues/37436.

### CODE REVIEW

**Comment by choidongyeon** [comment]
Note: flake8 is failing because of lint errors in ```test/jit/test_python_ir.py```.

File: torch/nn/modules/dropout.py [hunk 1]
```diff
@@ -181,6 +181,42 @@ def forward(self, input):
 
 
 class FeatureAlphaDropout(_DropoutNd):
+    r"""Randomly zero out entire channels (a channel is a 3D feature map,
+    e.g., the :math:`j`-th channel of the :math:`i`-th sample in the
+    batched input is a 3D tensor :math:`\text{input}[i, j]`).
+    Each channel will be zeroed out independently on every forward call with
```
Comment (SsnL) [inline torch/nn/modules/dropout.py:187]: Could you mention that it is independent for each sample?

File: torch/nn/modules/dropout.py [hunk 2]
```diff
@@ -181,6 +181,42 @@ def forward(self, input):
 
 
 class FeatureAlphaDropout(_DropoutNd):
+    r"""Randomly zero out entire channels (a channel is a 3D feature map,
+    e.g., the :math:`j`-th channel of the :math:`i`-th sample in the
+    batched input is a 3D tensor :math:`\text{input}[i, j]`).
```
Comment (SsnL) [inline torch/nn/modules/dropout.py:186]: It's not necessarily a 3D tensor, right? If the input are images, it is 2D.

File: torch/nn/functional.py [hunk 3]
```diff
@@ -1014,6 +1014,24 @@ def dropout3d(input, p=0.5, training=True, inplace=False):
...
+    activations to zero, as in regular Dropout, the activations are set to the 
+    negative saturation value of the SELU activation function.
+
+    Each element will be masked independently on every forward call with
+    probability :attr:`p` using samples from a Bernoulli distribution.
+    The elements to be masked are randomized on every forward call, and scaled
+    and shifted to maintain zero mean and unit variance.
+
+
+    See :class:`~torch.nn.AlphaDropout` for details.
```
Comment (SsnL) [inline torch/nn/functional.py:1028]: FeatureAlphaDropout?

File: torch/nn/functional.py [hunk 4]
```diff
@@ -1014,6 +1014,24 @@ def dropout3d(input, p=0.5, training=True, inplace=False):
 
 def feature_alpha_dropout(input, p=0.5, training=False, inplace=False):
     # type: (Tensor, float, bool, bool) -> Tensor
+    r"""
+    Randomly masks some of the elements of the input tensor. Instead of setting
```
Comment (SsnL) [inline torch/nn/functional.py:1018]: Mention that it wipes out entire channels?

File: torch/nn/modules/dropout.py [hunk 5]
```diff
@@ -181,6 +181,42 @@ def forward(self, input):
 
 
 class FeatureAlphaDropout(_DropoutNd):
+    r"""Randomly zero out entire channels (a channel is a 3D feature map,
```
Comment (SsnL) [inline torch/nn/modules/dropout.py:184]: This doesn't describe the Alpha part. Also need ref to the SELU paper.

**Review by ssnl — CHANGES_REQUESTED** [review]

File: torch/nn/functional.py (hunk 4 above)
Comment (choidongyeon) [reply torch/nn/functional.py:1018]: Addressed in next update.

File: torch/nn/functional.py (hunk 3 above)
Comment (choidongyeon) [reply torch/nn/functional.py:1028]: Addressed in next update.

File: torch/nn/modules/dropout.py (hunk 2 above)
Comment (choidongyeon) [reply torch/nn/modules/dropout.py:186]: Addressed in next update.

File: torch/nn/modules/dropout.py (hunk 5 above)
Comment (choidongyeon) [reply torch/nn/modules/dropout.py:184]: Addressed in next update.

File: torch/nn/modules/dropout.py (hunk 1 above)
Comment (choidongyeon) [reply torch/nn/modules/dropout.py:187]: Addressed in next update.

**Comment by choidongyeon** [comment]
Hey @apaszke, could you let me know what you think? :)

File: torch/nn/modules/dropout.py [hunk 6]
```diff
@@ -181,6 +181,46 @@ def forward(self, input):
...
+
+    Each element will be masked independently for each sample on every forward 
+    call with probability :attr:`p` using samples from a Bernoulli distribution.
+    The elements to be masked are randomized on every forward call, and scaled
+    and shifted to maintain zero mean and unit variance.
+
+    Usually the input comes from :class:`nn.Conv3d` modules.
+
+    As described in the paper
+    `Efficient Object Localization Using Convolutional Networks`_ ,
```
Comment (SsnL) [inline torch/nn/modules/dropout.py:198]: this is missing reference

File: torch/nn/modules/dropout.py [hunk 7]
```diff
@@ -181,6 +181,46 @@ def forward(self, input):
...
+        - Input: :math:`(N, C, D, H, W)`
+        - Output: :math:`(N, C, D, H, W)` (same shape as input)
+
+    Examples::
+
+        >>> m = nn.Dropout3d(p=0.2)
+        >>> input = torch.randn(20, 16, 4, 32, 32)
+        >>> output = m(input)
+
+    .. _Self-Normalizing Neural Networks: https://arxiv.org/abs/1706.02515
```
Comment (SsnL) [inline torch/nn/modules/dropout.py:222]: you are not using this reference in the text above...

**Review by ssnl — COMMENTED** [review]
lgtm other than two minor points. please fix. thanks!

File: torch/nn/modules/dropout.py (hunk 6 above)
Comment (choidongyeon) [reply torch/nn/modules/dropout.py]: Addressed in new update.

File: torch/nn/modules/dropout.py (hunk 7 above)
Comment (choidongyeon) [reply torch/nn/modules/dropout.py:221]: Addressed in new update.

**Comment by choidongyeon** [comment]
> lgtm other than two minor points. please fix. thanks!

Thanks for taking another look! Just updated the PR. Hopefully I got it right this time. :)

File: torch/nn/modules/dropout.py [hunk 8]
```diff
@@ -181,6 +181,45 @@ def forward(self, input):
...
+    to the negative saturation value of the SELU activation function.
+
+    Each element will be masked independently for each sample on every forward 
+    call with probability :attr:`p` using samples from a Bernoulli distribution.
+    The elements to be masked are randomized on every forward call, and scaled
+    and shifted to maintain zero mean and unit variance.
+
+    Usually the input comes from :class:`nn.Conv3d` modules.
+
+    As described in the paper `Self-Normalizing Neural Networks`_ ,
```
Comment (SsnL) [inline torch/nn/modules/dropout.py:197]: What this paragraph talks about is in the other paper, which you should keep, and add a reference (url) below.

In addition, you should add the reference to Self-Normalizing Neural Networks near the mention of SELU.

Comment (choidongyeon) [reply torch/nn/modules/dropout.py]: I see, I think I understand what you are saying - you're saying that two papers should be referenced. The first is "Self-Normalizing Neural Networks," which is to be referenced in the first paragraph re: SELU. The second is "Efficient Object Localization Using Convolutional Networks," which should be referenced here.

Just updated the PR based on this.

**Review by ssnl — APPROVED** [review]
Thank you! :)

**Review by ssnl — CHANGES_REQUESTED** [review]
oh you need to fix the submodule change

**Comment by choidongyeon** [comment]
> oh you need to fix the submodule change

Hey, what does this mean?

**Comment by SsnL** [comment]
no worries, i fixed it. 

@pytorchbot merge this please

**Review by ssnl — APPROVED** [review]

**Comment by choidongyeon** [comment]
@apaszke could you please take a look at this so we can close this out? :)

**Comment by apaszke** [comment]
@choidongyeon sorry, but I'm not at FB so I can't speed up merging this PR. @soumith?

**Review by zou3519 — CHANGES_REQUESTED** [review]
Requesting changes b/c of test failure

**Comment by choidongyeon** [comment]
Thanks for pointing that out @zou3519 !

**Comment by zou3519** [comment]
@choidongyeon I'm seeing some merge conflicts in the tests. This is usually caused by the branch being too far behind master (it look like it's around two days old, which really isn't that old...). Could you rebase onto master and force push to the branch?

After a rebase and the tests run I think this should be good to go. I can handle merging this when it's ready.

**Review by zou3519 — APPROVED** [review]

**Comment by choidongyeon** [comment]
@zou3519 Oh oops I merged master instead of rebase, hope that's okay.. Thanks for taking a look!

File: torch/nn/functional.py [hunk 9]
```diff
@@ -1051,6 +1051,25 @@ def dropout3d(input, p=0.5, training=True, inplace=False):
...
+
+    Each element will be masked independently on every forward call with
+    probability :attr:`p` using samples from a Bernoulli distribution.
+    The elements to be masked are randomized on every forward call, and scaled
+    and shifted to maintain zero mean and unit variance.
+
+    See :class:`~torch.nn.FeatureAlphaDropout` for details.
+
+    Args:
+        p: dropout probability of a channel to be zeroed.Default: 0.5
```
Comment (zou3519) [inline torch/nn/functional.py:1069]: nit: Missing space betwen the period and default "zeroed.Default"

File: torch/nn/modules/dropout.py [hunk 10]
```diff
@@ -181,6 +181,49 @@ def forward(self, input):
...
+    if adjacent pixels within feature maps are strongly correlated
+    (as is normally the case in early convolution layers) then i.i.d. dropout
+    will not regularize the activations and will otherwise just result
+    in an effective learning rate decrease.
+
+    In this case, :func:`nn.Dropout3d` will help promote independence between
+    feature maps and should be used instead.
+
+    Args:
+        p (float, optional): probability of an element to be zeroed.
```
Comment (zou3519) [inline torch/nn/modules/dropout.py:209]: nit: Should state the default here

File: torch/nn/modules/dropout.py [hunk 11]
```diff
@@ -181,6 +181,49 @@ def forward(self, input):
...
+        inplace (bool, optional): If set to ``True``, will do this operation
+            in-place
+
+    Shape:
+        - Input: :math:`(N, C, D, H, W)`
+        - Output: :math:`(N, C, D, H, W)` (same shape as input)
+
+    Examples::
+
+        >>> m = nn.Dropout3d(p=0.2)
```
Comment (zou3519) [inline torch/nn/modules/dropout.py:219]: Shouldn't this be FeatureAlphaDropout?

File: torch/nn/modules/dropout.py [hunk 12]
```diff
@@ -181,6 +181,49 @@ def forward(self, input):
...
+    will not regularize the activations and will otherwise just result
+    in an effective learning rate decrease.
+
+    In this case, :func:`nn.Dropout3d` will help promote independence between
+    feature maps and should be used instead.
+
+    Args:
+        p (float, optional): probability of an element to be zeroed.
+        inplace (bool, optional): If set to ``True``, will do this operation
+            in-place
```
Comment (zou3519) [inline torch/nn/modules/dropout.py:211]: Default should be stated here as well

**Comment by zou3519** [comment]
@choidongyeon yeah, that's fine. I had some quick comments while reading through the docs; sorry I should have done the read through when I requested changes initially.

File: torch/nn/modules/dropout.py [hunk 13]
```diff
@@ -181,6 +181,49 @@ def forward(self, input):
...
+
+    Shape:
+        - Input: :math:`(N, C, D, H, W)`
+        - Output: :math:`(N, C, D, H, W)` (same shape as input)
+
+    Examples::
+
+        >>> m = nn.FeatireAlphaDropout(p=0.2)
```
Comment (zou3519) [inline torch/nn/modules/dropout.py:219]: "Featire" -> "Feature"

**Review by zou3519 — APPROVED** [review]
lgtm, let's wait for the tests to pass

**Comment by vadimkantorov** [comment]
Somehow, it didn't work. Docs are not available online:
image

**Comment by vadimkantorov** [comment]
Probably because it misses `automodule` directive
